backtracking/n_queens.cpp

```diff
@@ -1,63 +1,123 @@
+ *
+ */
 #include <iostream>
-#define N 4
-using namespace std;
-
-void printSolution(int board[N][N]) {
-    cout << "\n";
-    for (int i = 0; i < N; i++) {
-        for (int j = 0; j < N; j++) cout << "" << board[i][j];
-        cout << "\n";
+#include <array>
+
+/**
+ * @namespace backtracking
+ * @brief Backtracking algorithms
+ */
+namespace backtracking {
+/**
+ * Utility function to print matrix
+ * @tparam n number of matrix size
+ * @param board matrix where numbers are saved
+ */
+template <size_t n>
+void printSolution(const std::array<std::array<int, n>, n> &board) {
+  std::cout << "\n";
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < n; j++) {
+      std::cout << "" << board[i][j];
     }
+    std::cout << "\n";
+  }
 }
 
-bool isSafe(int board[N][N], int row, int col) {
-    int i, j;
+/**
+ * Check if a queen can be placed on matrix
+ * @tparam n number of matrix size
+ * @param board matrix where numbers are saved
+ * @param row current index in rows
+ * @param col current index in columns
+ * @returns `true` if queen can be placed on matrix
+ * @returns `false` if queen can't be placed on matrix
+ */
+template <size_t n>
+bool isSafe(const std::array<std::array<int, n>, n> &board, const int &row,
+            const int &col) {
+  int i = 0, j = 0;
 
-    /* Check this row on left side */
-    for (i = 0; i < col; i++)
-        if (board[row][i])
-            return false;
+  // Check this row on left side
+  for (i = 0; i < col; i++) {
+    if (board[row][i]) {
+      return false;
+    }
+  }
 
-    /* Check upper diagonal on left side */
-    for (i = row, j = col; i >= 0 && j >= 0; i--, j--)
-        if (board[i][j])
-            return false;
+  // Check upper diagonal on left side
+  for (i = row, j = col; i >= 0 && j >= 0; i--, j--) {
+    if (board[i][j]) {
+      return false;
+    }
+  }
+  // Check lower diagonal on left side
+  for (i = row, j = col; j >= 0 && i < n; i++, j--) {
+    if (board[i][j]) {
+      return false;
+    }
+  }
+  return true;
+}
 
-    /* Check lower diagonal on left side */
-    for (i = row, j = col; j >= 0 && i < N; i++, j--)
-        if (board[i][j])
-            return false;
+/**
+ * Solve n queens problem
+ * @tparam n number of matrix size
+ * @param board matrix where numbers are saved
+ * @param col current index in columns
+ */
+template <size_t n>
+void solveNQ(std::array<std::array<int, n>, n> board, const int &col) {
+  if (col >= n) {
+    printSolution<n>(board);
+    return;
+  }
 
-    return true;
-}
+  // Consider this column and try placing
+  // this queen in all rows one by one
+  for (int i = 0; i < n; i++) {
+    // Check if queen can be placed
+    // on board[i][col]
+    if (isSafe<n>(board, i, col)) {
+      // Place this queen in matrix
+      board[i][col] = 1;
 
-void solveNQ(int board[N][N], int col) {
-    if (col >= N) {
-        printSolution(board);
-        return;
-    }
+      // Recur to place rest of the queens
```

Comment:
A minor correction
**Recursive call** to place rest of the queens

Reply:
Fixed, thank you.